### PR TITLE
[v6r15] VOMS2CSAgent

### PR DESCRIPTION
**Pull request by atsareg**
New VOMS2CSAgent to synchronize user information in VOMS and in the DIRAC CS Registry. Unlike UsersAndGroupAgent, this one works with multiple VOs.

### CODE REVIEW

File: ConfigurationSystem/Agent/VOMS2CSAgent.py [hunk 1]
```diff
@@ -0,0 +1,323 @@
...
+from DIRAC                                             import S_OK, S_ERROR, gConfig
+from DIRAC.ConfigurationSystem.Client.Helpers.Registry import getVOOption, getUserOption, \
+                                                              getVOMSRoleGroupMapping, getUsersInVO, \
+                                                              getAllUsers
+from DIRAC.Core.Utilities.Proxy                        import executeWithUserProxy
+
+class VOMS2CSAgent( AgentModule ):
+
+  def initialize( self ):
+    self.am_setOption( "PollingTime", 3600 * 6 ) # Every 6 hours
```
Comment (fstagni) [inline ConfigurationSystem/Agent/VOMS2CSAgent.py]: Why here and not in ConfigTemplate.cfg? If it's here is hidden to the administrator.

**Comment by fstagni** [comment]
Question: is this agent a replacement, or an alternative to UsersAndGroups?

File: Core/Security/VOMSService.py [hunk 2]
```diff
@@ -1,46 +1,72 @@
...
+      for info in entry:
+
+        try:
+          entryData[ info[0] ] = str( info[1] )
+        except:
+          pass
+      data.append( entryData )
   return data
 
 class VOMSService:
```
Comment (fstagni) [inline Core/Security/VOMSService.py]: inherit from object

File: Core/Security/VOMSService.py [hunk 3]
```diff
@@ -1,46 +1,72 @@
...
 
-  def __init__( self, adminUrl = False, attributesUrl = False, certificatesUrl = False ):
+  def __init__( self, vo, adminUrl = False, attributesUrl = False, certificatesUrl = False ):
```
Comment (fstagni) [inline Core/Security/VOMSService.py]: This change in signature is a bit dangerous. The class is used also in extensions.

File: Core/Security/VOMSService.py [hunk 4]
```diff
@@ -105,3 +131,41 @@ def attGetUserNickname( self, dn, ca ):
...
+    members = result['Value']
+
+    result = self.admListRoles()
+    if not result['OK']:
+      return result
+    roles = result['Value']
+
+    roleMembers = {}
+    for role in roles:
+      result = self.admListUsersWithRole( '/%s' % self.vomsVO, role )
```
Comment (fstagni) [inline Core/Security/VOMSService.py]: double checking: does role starts with "Role=", like in https://github.com/DIRACGrid/DIRAC/blob/integration/ConfigurationSystem/Agent/UsersAndGroups.py#L262 ? Also, at same location, the first parameter to admListUsersWithRole is passed in without "/"

File: ConfigurationSystem/Agent/VOMS2CSAgent.py (hunk 1 above)
Comment (atsareg) [inline ConfigurationSystem/Agent/VOMS2CSAgent.py]: Agreed. UserAndGroupAgent legacy

**Comment by atsareg** [comment]
@fstagni : In most cases this will be a replacement, in this respect it is more or less equivalent to the UserAndGroupAgent. For LHCb it should be decided what to do with the FC user registration. If nothing ( because DFC adds users automatically ) then it can just switch to VOMS2CSAgent in order not to have its own.

File: Core/Security/VOMSService.py (hunk 4 above)
Comment (atsareg) [inline Core/Security/VOMSService.py]: I will check, this is not a new functionality.

File: Core/Security/VOMSService.py (hunk 3 above)
Comment (atsareg) [inline Core/Security/VOMSService.py]: True. However, the VO is now a
      crucial attribute of objects of this class. I think, I will leave
      only the vo argument as a mandatory one and will make the
      discovery of connection details
      inside this class.
      
      On 28/10/15 11:24, fstagni wrote:
    
    
      In Core/Security/VOMSService.py:
      >    return data
>  
>  class VOMSService:
>  
> -  def __init__( self, adminUrl = False, attributesUrl = False, certificatesUrl = False ):
> +  def __init__( self, vo, adminUrl = False, attributesUrl = False, certificatesUrl = False ):

      This change in signature is a bit dangerous. The class is used
        also in extensions.
      —
        Reply to this email directly or view
          it on GitHub.

File: Core/Security/VOMSService.py [hunk 5]
```diff
@@ -1,46 +1,78 @@
...
+    self.vo = vo
+    self.vomsVO = getVOOption( vo, "VOMSName" )
+    if not self.vomsVO:
+      raise Exception( "Can not get VOMS name for VO %s" % vo )
     self.__soapClients = {}
     for key, url in ( ( 'Admin', adminUrl ), ( 'Attributes', attributesUrl ), ( 'Certificates', certificatesUrl ) ):
+      urls = []
       if not url:
-        url = gConfig.getValue( "/Registry/VOMS/URLs/VOMS%s" % key, "" )
+        url = gConfig.getValue( "/Registry/VO/%s/VOMSServices/VOMS%s" % ( self.vo, key ), "" )
```
Comment (fstagni) [inline Core/Security/VOMSService.py]: In case of a single VO installation, this is not correct.

File: Core/Security/VOMSService.py [hunk 6]
```diff
@@ -1,46 +1,78 @@
...
-      while retries:
-        retries -= 1
-        try:
-          client = getSOAPClient( "%s?wsdl" % url )
-          client.set_options(headers={"X-VOMS-CSRF-GUARD":"1"})
-          self.__soapClients[ key ] = client
-          break
-        except Exception:
-          if retries:
+        result = gConfig.getSections( '/Registry/VO/%s/VOMSServers' % self.vo )
```
Comment (fstagni) [inline Core/Security/VOMSService.py]: And this one neither.

File: Core/Security/VOMSService.py (hunk 5 above)
Comment (atsareg) [inline Core/Security/VOMSService.py]: Why ? It is correct, look at the LHCb CS, for example. The VOMS section is obsoleted. To keep the same arrangement for single and multi VO installations

Comment (fstagni) [inline Core/Security/VOMSService.py]: ah, you are indeed right.

**Comment by fstagni** [comment]
We need announcement of this new agent in https://github.com/DIRACGrid/DIRAC/wiki/DIRAC-v6r15

**Comment by chaen** [comment]
Just to comment on "because DFC adds users automatically": it adds the users, but an admin still needs to create the home directory because of the permissions we have set

**Pull request merged**